Sources/Testing/Traits/Trait.swift

```diff
@@ -81,6 +81,41 @@ public protocol SuiteTrait: Trait {
   var isRecursive: Bool { get }
 }
 
+/// A protocol extending ``Trait`` that offers an additional customization point
+/// for trait authors to execute code before and after each test function (if
+/// added to the traits of a test function), or before and after each test suite
+/// (if added to the traits of a test suite).
+public protocol CustomExecutionTrait: Trait {
+
+  /// Execute a function with the effects of this trait applied.
+  ///
+  /// - Parameters:
+  ///   - function: The function to perform. If `test` represents a test suite,
+  ///     this function encapsulates running all the tests in that suite. If
+  ///     `test` represents a test function, this function is the body of that
+  ///     test function (including all cases if it is parameterized.)
+  ///   - test: The test under which `function` is being performed.
+  ///   - testCase: The testCase under which `function` is being performed.
```

Comment:
Can the wording about the `testCase` parameter include "if any", to clarify that it's Optional? And then add a sentence explaining what circumstances it would be nil (a suite)